Provide an initial implementation of `text-wrap: balance`

RWDavid opened the pull request:
#### 2d299a2dd8bbf0374011075326ffb8c83433cc4a
<pre>
Provide an initial implementation of `text-wrap: balance`
<a href="https://bugs.webkit.org/show_bug.cgi?id=249840">https://bugs.webkit.org/show_bug.cgi?id=249840</a>
rdar://problem/103663513

Reviewed by Alan Baradlay.

This patch implements `text-wrap: balance` in IFC. It uses a Knuth-Plass
style dynamic programming algorithm to minimize the variance of all the
line lengths. There is no line limit imposed on this implementation.

Ideally, the act of balancing inline content will use the same number of
lines as if the inline content was laid out via `text-wrap: wrap`. However,
adhering to this ideal is expensive, and not caring about this ideal will
allow us to use a more efficient algorithm. Since inline content spanning
many lines is less likely to care about the exact preservation of vertical
space (i.e. number of lines used), we ignore this ideal number of lines
requirement beyond a certain line count (tentatively set to 12 lines for now).

There are two implementations of the Knuth-Plass style algorithm:
    (1) one which attempts to preserve the number of lines used, and
    (2) one that ignores that requirement.
If we let N denote the number of break opportunities, let L denote the
number of lines used, and let K denote the maximum number of inline items
that can fit in a single line, then algorithm (1) has a time complexity of
O(N * L * K), while algorithm (2) has a time complexity of O(N * K).

We use algorithm (1) when we want to preserve the number of lines used,
and we use algorithm (2) when we want to prioritize performance (i.e. when
inline content spans many lines). It is worth noting that algorithm (2)
will often also preserve the number of lines used. In the case that it
does not, the number of lines will differ only by a small amount.

This is an initial implementation, and not all features are supported.
Unsupported features include (and are not limited to):
- first-line styling
- indentation
- white-space-collapse
- tabs
- hyphens
- floats (including initial-letter)

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.cpp: Added.
(WebCore::Layout::computeBreakOpportunities):
(WebCore::Layout::computeCost):
(WebCore::Layout::InlineContentBalancer::InlineContentBalancer):
(WebCore::Layout::InlineContentBalancer::initialize):
(WebCore::Layout::InlineContentBalancer::computeBalanceConstraints):
(WebCore::Layout::InlineContentBalancer::balanceRangeWithLineRequirement):
(WebCore::Layout::InlineContentBalancer::balanceRangeWithNoLineRequirement):
(WebCore::Layout::InlineContentBalancer::inlineItemWidth const):
(WebCore::Layout::InlineContentBalancer::shouldTrimLeading const):
(WebCore::Layout::InlineContentBalancer::shouldTrimTrailing const):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::SlidingWidth):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::width):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::advanceStart):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::advanceStartTo):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::advanceEnd):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::advanceEndTo):
* Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.h: Added.
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::InlineFormattingContext::layoutInFlowAndFloatContent):
* Source/WebCore/layout/formattingContexts/inline/TextOnlySimpleLineBuilder.cpp:
(WebCore::Layout::TextOnlySimpleLineBuilder::isEligibleForSimplifiedTextOnlyInlineLayout):
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
(WebCore::LayoutIntegration::shouldInvalidateLineLayoutPathAfterChangeFor):

Canonical link: <a href="https://commits.webkit.org/266980@main">https://commits.webkit.org/266980@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26a467cfedef116125d90805d4d91be17f02f26c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15284 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15587 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15949 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17040 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14348 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15424 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18104 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15687 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16954 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15467 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15896 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12994 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17773 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13175 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13785 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20739 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14256 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13952 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17208 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14525 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/12307 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13796 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3672 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18143 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14359 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay reviewed (approved):
TextOnlyLineBuilder name is a bit misleading and I am going to rename it to TextOnlySimpleLineBuilder. It is supposed to process the most common text content in a very cheap manner. While adding text-wrap: balance is certainly not a perf hit, I'd rather keep it super simple for now.

RWDavid committed:
Provide an initial implementation of `text-wrap: balance`
https://bugs.webkit.org/show_bug.cgi?id=249840
rdar://problem/103663513

Reviewed by Alan Baradlay.

This patch implements `text-wrap: balance` in IFC. It uses a Knuth-Plass
style dynamic programming algorithm to minimize the variance of all the
line lengths. There is no line limit imposed on this implementation.

Ideally, the act of balancing inline content will use the same number of
lines as if the inline content was laid out via `text-wrap: wrap`. However,
adhering to this ideal is expensive, and not caring about this ideal will
allow us to use a more efficient algorithm. Since inline content spanning
many lines is less likely to care about the exact preservation of vertical
space (i.e. number of lines used), we ignore this ideal number of lines
requirement beyond a certain line count (tentatively set to 12 lines for now).

There are two implementations of the Knuth-Plass style algorithm:
    (1) one which attempts to preserve the number of lines used, and
    (2) one that ignores that requirement.
If we let N denote the number of break opportunities, let L denote the
number of lines used, and let K denote the maximum number of inline items
that can fit in a single line, then algorithm (1) has a time complexity of
O(N * L * K), while algorithm (2) has a time complexity of O(N * K).

We use algorithm (1) when we want to preserve the number of lines used,
and we use algorithm (2) when we want to prioritize performance (i.e. when
inline content spans many lines). It is worth noting that algorithm (2)
will often also preserve the number of lines used. In the case that it
does not, the number of lines will differ only by a small amount.

This is an initial implementation, and not all features are supported.
Unsupported features include (and are not limited to):
- first-line styling
- indentation
- white-space-collapse
- tabs
- hyphens
- floats (including initial-letter)

* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.cpp: Added.
(WebCore::Layout::computeBreakOpportunities):
(WebCore::Layout::computeCost):
(WebCore::Layout::InlineContentBalancer::InlineContentBalancer):
(WebCore::Layout::InlineContentBalancer::initialize):
(WebCore::Layout::InlineContentBalancer::computeBalanceConstraints):
(WebCore::Layout::InlineContentBalancer::balanceRangeWithLineRequirement):
(WebCore::Layout::InlineContentBalancer::balanceRangeWithNoLineRequirement):
(WebCore::Layout::InlineContentBalancer::inlineItemWidth const):
(WebCore::Layout::InlineContentBalancer::shouldTrimLeading const):
(WebCore::Layout::InlineContentBalancer::shouldTrimTrailing const):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::SlidingWidth):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::width):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::advanceStart):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::advanceStartTo):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::advanceEnd):
(WebCore::Layout::InlineContentBalancer::SlidingWidth::advanceEndTo):
* Source/WebCore/layout/formattingContexts/inline/InlineContentBalancer.h: Added.
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::InlineFormattingContext::layoutInFlowAndFloatContent):
* Source/WebCore/layout/formattingContexts/inline/TextOnlySimpleLineBuilder.cpp:
(WebCore::Layout::TextOnlySimpleLineBuilder::isEligibleForSimplifiedTextOnlyInlineLayout):
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:
(WebCore::LayoutIntegration::shouldInvalidateLineLayoutPathAfterChangeFor):

Canonical link: https://commits.webkit.org/266980@main
